stdlib/domain.ml

```diff
@@ -1,11 +1,90 @@
-type t = int
-external spawn : (unit -> unit) -> t = "caml_domain_spawn"
+module Raw = struct
+  (* Low-level primitives provided by the runtime *)
+  type t = private int
+  external critical_adjust : int -> unit
+    = "caml_ml_domain_critical_section"
+  external interrupt : t -> unit
+    = "caml_ml_domain_interrupt"
+  external wait : unit -> unit = "caml_ml_domain_yield"
+  external spawn : (unit -> unit) -> t
+    = "caml_domain_spawn"
+  external self : unit -> t
+    = "caml_ml_domain_id"
+end
+
+module Sync = struct
+  let critical_section f =
+    Raw.critical_adjust (+1);
+    match f () with
+    | x -> Raw.critical_adjust (-1); x
+    | exception ex -> Raw.critical_adjust (-1); raise ex
```

Comment:
Ok. That makes sense. Effect handlers that return immediately should be fine. If you drop the continuation, then the system will make progress after GC. The only tricky case would be when the continuation is stashed away in data structure for later resumption. At this point, any notifications will block until the continuation resumes. There is a risk of deadlock here.